nuxeo-core/nuxeo-core-mongodb/src/main/java/org/nuxeo/ecm/core/mongodb/kv/MongoDBKeyValueStore.java

```diff
@@ -0,0 +1,276 @@
+import com.mongodb.client.model.IndexOptions;
+import com.mongodb.client.model.UpdateOptions;
+import com.mongodb.client.result.DeleteResult;
+import com.mongodb.client.result.UpdateResult;
+
+/**
+ * MongoDB implementation of a Key/Value Store Provider.
+ * <p>
+ * The following configuration properties are available:
+ * <ul>
+ * <li>collection: the MongoDB collection prefix to use, the default is "kv". This will be followed by the Store name.
+ * </ul>
+ *
+ * @since 9.3
+ */
+public class MongoDBKeyValueStore implements KeyValueStoreProvider {
+
+    private static final Log log = LogFactory.getLog(MongoDBKeyValueStore.class);
+
+    public static final String KEYVALUE_CONNECTION_ID = "keyvalue";
+
+    public static final String COLLECTION_PROP = "collection";
+
+    public static final String COLLECTION_DEFAULT = "kv";
+
+    public static final String ID_KEY = "_id";
+
+    public static final String VALUE_KEY = "v";
+
+    public static final String TTL_KEY = "ttl";
+
+    public static final Double ONE = Double.valueOf(1);
+
+    protected String name;
+
+    protected MongoCollection<Document> coll;
+
+    protected static ThreadLocal<CharsetDecoder> UTF_8_DECODERS = ThreadLocal.withInitial(
+            () -> UTF_8.newDecoder().onMalformedInput(CodingErrorAction.REPORT).onUnmappableCharacter(
+                    CodingErrorAction.REPORT));
+
+    @Override
+    public void initialize(KeyValueStoreDescriptor descriptor) {
+        name = descriptor.name;
+        Map<String, String> properties = descriptor.getProperties();
+        // find which collection prefix to use
+        String collectionName = properties.get(COLLECTION_PROP);
+        if (StringUtils.isBlank(collectionName)) {
+            collectionName = COLLECTION_DEFAULT;
+        }
+        collectionName += "." + name;
+        // get a connection to MongoDB
+        MongoDBConnectionService mongoService = Framework.getService(MongoDBConnectionService.class);
+        MongoDatabase database = mongoService.getDatabase(KEYVALUE_CONNECTION_ID);
+        coll = database.getCollection(collectionName);
+        // make sure TTL works by creating the appropriate index
+        IndexOptions indexOptions = new IndexOptions().expireAfter(Long.valueOf(0), TimeUnit.SECONDS);
+        coll.createIndex(new Document(TTL_KEY, ONE), indexOptions);
+    }
+
+    @Override
+    public void close() {
+        if (coll != null) {
+            coll = null;
+        }
+    }
+
+    @Override
+    public void clear() {
+        if (log.isTraceEnabled()) {
+            log.trace("MongoDB: CLEAR");
+        }
+        coll.deleteMany(new Document());
+    }
+
+    // if possible, store the bytes as a UTF-8 string
+    protected static Object toStorage(byte[] bytes) {
+        try {
+            return UTF_8_DECODERS.get().decode(ByteBuffer.wrap(bytes)).toString();
+        } catch (CharacterCodingException e) {
+            // could not decode as UTF-8, use a binary
```

Comment:
Why do we want utf8 instead of binary, for human readability ? here the cost is that normal operation go through exception which perf impacts

Reply:
Yes for human readability and the possibility to debug/modify things easily at the database level. I don't like control flow through exceptions either but I didn't find any other way to do it.

Regarding perfs yes this has an impact but it should be negligible as K/V Store is not really heavily used today :)

An other option I was thinking about was to provide an API to K/V Store that takes a `String` instead of a `byte[]`. And maybe another with some form of JSON to allow direct mapping in MongoDB at least.

Reply:
ok so let see the usage and add new methods later